### ISSUE TITLE
The model output differs drastically for curated classical texts. I ran this as ONNX after conversion from HF on .Net platform.

### ISSUE THREAD

**Issue by mediabuff** [sebastian-nehrdich/byt5-sanskrit-analyzers#5]
```
======================================================================
          CLASSICAL SANSKRIT GOLD-STANDARD BENCHMARK RUNNER
======================================================================
Target Text Layer: Valmiki Ramayana & Bhagavad Gita Ground Truth
Corpus Volume:    4 Curated Syntactic Units

──────────────────────────────────────────────────────────────────────
Source Location: Valmiki Ramayana - Bala Kanda 1.1.1
[Raw Input]:     ikṣvākuvaṃśaprabhavo rāmo nāma janaiḥ śrutaḥ
  ├─ Phonetic Segment:   "ikṣvākuvaṃśaprabhavo rāmo nāma janaiḥ śrutaḥ"
  │  └─ Ground Truth:    "ikṣvāku vaṃśa prabhavo rāmo nāma janaiḥ śrutaḥ" -> ❌ MISALIGN
  ├─ Reconstructed Base: "ikṣvākuvaṃśaprabhavo rāmo nāma janaiḥ śrutaḥ"
  │  └─ Gold Padapāṭha:  "ikṣvāku vaṃśa prabhavaḥ rāmaḥ nāma janaiḥ śrutaḥ" -> ❌ DEVIATION
  └─ Metrics: Latency=1666.75ms | CER=12.5%
──────────────────────────────────────────────────────────────────────
Source Location: Valmiki Ramayana - Bala Kanda 1.1.8
[Raw Input]:     buddhimānnītīmāñchālyo vāg mī ruciracauditaḥ
  ├─ Phonetic Segment:   "buddhimānnītīmāñchālyo vāg mī ruciracauditaḥ"
  │  └─ Ground Truth:    "buddhimān nītīmān chālyo vāg mī rucira cauditaḥ" -> ❌ MISALIGN
  ├─ Reconstructed Base: "buddhimānnītīmāñchālyo vāg mī ruciracauditaḥ"
  │  └─ Gold Padapāṭha:  "buddhimān nītimān śālyaḥ vāk mī rucira coditaḥ" -> ❌ DEVIATION
  └─ Metrics: Latency=1367.62ms | CER=23.9%
──────────────────────────────────────────────────────────────────────
Source Location: Bhagavad Gita - Chapter 1.1
[Raw Input]:     dharmakṣetre kurukṣetre samavetā yuyutsavaḥ
  ├─ Phonetic Segment:   "dharmakṣetre kurukṣetre samavetā yuyutsavaḥ"
  │  └─ Ground Truth:    "dharmakṣetre kurukṣetre samavetā yuyutsavaḥ" -> ✔ MATCH
  ├─ Reconstructed Base: "dharmakṣetre kurukṣetre samavetā yuyutsavaḥ"
  │  └─ Gold Padapāṭha:  "dharma kṣetre kuru kṣetre samavetāḥ yuyutsavaḥ" -> ❌ DEVIATION
  └─ Metrics: Latency=2669.99ms | CER=6.5%
──────────────────────────────────────────────────────────────────────
Source Location: Bhagavad Gita - Chapter 2.22
[Raw Input]:     vāsāṃsi jīrṇāni yathā vihāya navāni gṛhṇāti naro'parāṇi
  ├─ Phonetic Segment:   "ṃ jīrṇāni yathā vihāya navāni gṛhṇāti naro'parāṇi"
  │  └─ Ground Truth:    "vāsāṃsi jīrṇāni yathā vihāya navāni gṛhṇāti naro aparāṇi" -> ❌ MISALIGN
  ├─ Reconstructed Base: "ṃ jīrṇāni yathā vihāya navāni gṛhṇāti naro'parāṇi"
  │  └─ Gold Padapāṭha:  "vāsas jīrṇa yathā vihāya nava gṛhṇāti naraḥ aparāṇi" -> ❌ DEVIATION
  └─ Metrics: Latency=5871.27ms | CER=29.4%

======================================================================
                    NLP DEEP LEARNING METRIC REPORT
======================================================================
Total Test Sentences:       4
Phonetic Segmentation PM:   1 / 4 (25.00%)
Morphological Lemma PM:     0 / 4 (0.00%)
Average Character Error Rate (CER): 18.09%
Total Pipeline Processing Time:     11585.92 ms
=====================================================================
```

**Comment by mediabuff** [comment]
Notes from Google Gemini on my 2 implementations  using the model here.

***********************
Whether `AdaptiveSanskritSegmenter` performs better than a `ByT5` based segmenter depends on the architectural trade-off between **explicit linguistic control** and **latent sequence modeling**.

Since you have frozen development on your current implementation, here is an objective assessment of why one might be preferred over the other for a "Classic Corpus" task:

### 1\. ByT5SanskritSegmenter (The Latent Learner)

*   **Strengths:** `ByT5` is a character-level, byte-to-byte model. It is exceptionally robust at capturing the morphological nuances of Sanskrit—such as _sandhi_ rules and inflections—without needing hard-coded logic. It treats the corpus as a raw sequence, which often yields higher "naturalness" in output.
    
*   **Weaknesses:** As a pure sequence-to-sequence model, it suffers from **hallucination and repetition loops** in long-sequence generation. It lacks a "hard" mechanism to force the model to respect the input boundaries, which is why you see the _kurukṣetre_ stutter; the model "predicts" that the verse _must_ be repeated because its internal attention weights are saturated by the training data.
    

### 2\. AdaptiveSanskritSegmenter (The Controlled System)

*   **Strengths:** The `Adaptive` approach (your current code) is architected for **system-level constraint**. By manually managing the `encoder_attention_mask` and implementing logic to detect repetitive loops, you are essentially "guard-railing" the decoder. It is inherently better at _stability_ because it allows you to inject logic that is external to the model weights.
    
*   **Weaknesses:** The complexity of managing these masks and logits manually can introduce the bugs we encountered (e.g., `Missing Input` or `NaN` values). It is also susceptible to "context collapse" if the masking logic is too aggressive, as it can strip the model of the context it needs to correctly segment complex _sandhi_.
    

### Conclusion for Classical Corpus

*   **If your priority is accuracy on high-frequency verses:** `ByT5SanskritSegmenter` is typically superior because it captures the statistical likelihood of character transitions better than any manually tuned heuristic.
    
*   **If your priority is avoiding catastrophic failures (stuttering) in a production environment:** `AdaptiveSanskritSegmenter` is conceptually better, _provided the masking and penalty logic is calibrated precisely_.
    

The reason you are struggling now is that the `Adaptive` logic is fighting the `Decoder` model's internal confidence. In classical Sanskrit, certain verses are so over-represented in training data that the model’s internal probability for "what comes next" is almost absolute. The `Adaptive` segmenter is trying to nudge that probability with a small penalty (`-2.5f`), but the model's learned weights are often stronger than that nudge.

**Comment by mediabuff** [comment]
Whether rule/lexicon-driven models are "better" than data-driven models (like Large Language Models or sequence-to-sequence models) for Sanskrit segmentation is a subject of active debate in the field of computational linguistics. The answer depends entirely on your constraints regarding **precision, domain coverage, and computational cost.**

### 1\. Rule and Lexicon-Driven Models

These systems rely on formal grammars (typically Pāṇinian) and extensive dictionaries (like the _Monier-Williams_ or _Cologne Sanskrit Lexicon_).

*   **Strengths:**
    
    *   **Interpretability:** Since they follow specific grammatical rules (_sandhi_ and _samāsa_), you can trace exactly why a segmentation decision was made.
        
    *   **Precision:** They are less prone to "hallucinations" or generating non-existent words because they are constrained by the dictionary.
        
    *   **Zero-Shot Capability:** They can often segment rare or newly encountered words if the rules of morphology apply.
        
*   **Weaknesses:**
    
    *   **Complexity (The "Sanskrit Ambiguity" Problem):** Sanskrit is highly agglutinative. A single string of characters can have dozens of valid grammatical segmentations. Rule-driven systems often struggle to rank these, leading to a massive list of potential candidates rather than the "correct" one.
        
    *   **Fragility:** They often fail when faced with non-standard Sanskrit (e.g., epic, Buddhist, or late-period literature) that deviates from classical Pāṇinian norms.
        

### 2\. LLM/Data-Driven Models (e.g., ByT5, BERT-based)

These models learn statistical patterns of character/token distribution from massive corpora.

*   **Strengths:**
    
    *   **Context Sensitivity:** They are excellent at disambiguation. While a rule-based system sees many possible segmentations, a data-driven model understands which one is statistically most likely given the surrounding words.
        
    *   **Robustness to Noise:** They handle non-standard grammar, typos, and variations in historical orthography much better than rigid rule sets.
        
*   **Weaknesses:**
    
    *   **Lack of "Ground Truth" Logic:** They do not "know" grammar; they mimic it. If the training data was biased or lacked certain literary styles, the model will produce high-confidence errors.
        
    *   **The Repetition Trap:** As you have experienced with the _kurukṣetre_ loop, these models can become "over-confident" in specific verse patterns, leading to catastrophic failure in repetitive classical texts.
        

<img width="777" height="367" alt="Image" src="https://github.com/user-attachments/assets/1c146307-4770-4db1-87fd-37652f22e605" />

### The Modern Consensus: Hybrid Approaches

The industry standard is moving away from choosing one over the other. Most high-performing Sanskrit processing pipelines use a **Hybrid Approach**:

1.  **Candidate Generation (Rule-driven):** Use a lexicon-driven model to generate all _mathematically/grammatically possible_ segmentations.
    
2.  **Candidate Ranking (Data-driven):** Use a lightweight data-driven model (like a transformer) to score those candidates based on context and pick the most probable one.
    

This combines the **precision** of grammar with the **contextual awareness** of LLMs.

For your specific project, the difficulty you are facing with the `AdaptiveSanskritSegmenter` is essentially the result of using a data-driven model as if it were a rule-driven one. If you want to improve performance, you might consider keeping the data-driven decoder but using a constraint-based "beam search" that forces the model to only output segments that exist in a trusted lexicon.
